File: doctranslator/project.py
from __future__ import annotations

import re
from dataclasses import dataclass

from doctranslator.freezer import unfreeze_text
from doctranslator.ir import FormatSpan, TextNodeRef
from doctranslator.sentinels import CONTROL_TOKENS, control_tokens_from_text


_TOKEN_RE = re.compile(r"⟦[^⟧]+⟧")


@dataclass(frozen=True)
class SpanSlice:
    span: FormatSpan
    text: str


def _split_by_control_sequence(text: str) -> list[str]:
    if not text:
        return [""]
    pattern = re.compile("|".join(re.escape(tok) for tok in CONTROL_TOKENS))
    parts: list[str] = []
    pos = 0
    for m in pattern.finditer(text):
        parts.append(text[pos : m.start()])
        parts.append(m.group(0))
        pos = m.end()
    parts.append(text[pos:])
    return parts


def _unitize(text: str) -> list[str]:
    units: list[str] = []
    pos = 0
    for m in _TOKEN_RE.finditer(text):
        if m.start() > pos:
            units.extend(list(text[pos : m.start()]))
        units.append(m.group(0))
        pos = m.end()
    if pos < len(text):
        units.extend(list(text[pos:]))
    return units


def _count_plain_units(units: list[str]) -> int:
    return sum(1 for u in units if _TOKEN_RE.fullmatch(u) is None)


def _allocate_plain_counts(total_plain: int, weights: list[int]) -> list[int]:
    if not weights:
        return []
    if total_plain <= 0:
        return [0 for _ in weights]

    total_w = sum(weights)
    if total_w <= 0:
        base = total_plain // len(weights)
        out = [base for _ in weights]
        out[-1] += total_plain - sum(out)
        return out

    raw = [total_plain * w / total_w for w in weights]
    floored = [int(x) for x in raw]
    remain = total_plain - sum(floored)
    frac = sorted([(raw[i] - floored[i], i) for i in range(len(weights))], reverse=True)
    for k in range(remain):
        _, idx = frac[k % len(frac)]
        floored[idx] += 1
    return floored


def project_translation_to_spans(
    spans: list[FormatSpan], source_surface: str, target_surface: str, nt_map: dict[str, str]
) -> list[SpanSlice]:
    if control_tokens_from_text(source_surface) != control_tokens_from_text(target_surface):
        raise ValueError("Control token sequence mismatch")

    source_parts = _split_by_control_sequence(source_surface)
    target_parts = _split_by_control_sequence(target_surface)
    if len(source_parts) != len(target_parts):
        raise ValueError("Control token part count mismatch")

    span_slices: list[SpanSlice] = []
    span_idx = 0

    for src_part, tgt_part in zip(source_parts, target_parts, strict=True):
        if src_part in CONTROL_TOKENS:
            if tgt_part != src_part:
                raise ValueError("Control token mismatch")
            continue

        block_spans: list[FormatSpan] = []
        block_src_len = 0
        while span_idx < len(spans) and block_src_len < len(src_part):
            span = spans[span_idx]
            block_spans.append(span)
            block_src_len += len(span.source_text)
            span_idx += 1

        if not block_spans:
            if tgt_part.strip():
                raise ValueError("Translated text exists for empty source block")
            continue

        tgt_units = _unitize(tgt_part)
        total_plain = _count_plain_units(tgt_units)
        weights = [max(len(span.source_text), 1) for span in block_spans]
        desired = _allocate_plain_counts(total_plain, weights)

        slices_units: list[list[str]] = [[] for _ in block_spans]
        current_span = 0
        current_plain = 0

        for unit in tgt_units:
            slices_units[current_span].append(unit)
            if _TOKEN_RE.fullmatch(unit) is None:
                current_plain += 1
            if current_span < len(block_spans) - 1 and current_plain >= desired[current_span]:
                current_span += 1
                current_plain = 0

        for span, units in zip(block_spans, slices_units, strict=True):
            span_text_frozen = "".join(units)
            span_text = unfreeze_text(span_text_frozen, nt_map)
            span_slices.append(SpanSlice(span=span, text=span_text))

    if span_idx != len(spans):
        remaining = spans[span_idx:]
        if any(s.source_text.strip() for s in remaining):
            raise ValueError("Span coverage mismatch")
        for span in remaining:
            span_slices.append(SpanSlice(span=span, text=""))

    return span_slices


def distribute_span_text_to_nodes(span: FormatSpan, text: str) -> list[tuple[TextNodeRef, str]]:
    if not span.node_refs:
        return []
    if len(span.node_refs) == 1:
        return [(span.node_refs[0], text)]

    weights = [max(len(n.original_text), 1) for n in span.node_refs]
    units = list(text)
    total = len(units)
    desired = _allocate_plain_counts(total, weights)
    out: list[tuple[TextNodeRef, str]] = []
    idx = 0
    for node_ref, count in zip(span.node_refs, desired, strict=True):
        piece = "".join(units[idx : idx + count])
        idx += count
        out.append((node_ref, piece))
    if idx < len(units):
        last_ref, last_text = out[-1]
        out[-1] = (last_ref, last_text + "".join(units[idx:]))
    return out

File: src/docx/project.rs
use anyhow::anyhow;

use crate::freezer::unfreeze_text;
use crate::ir::{FormatSpan, TextNodeRef};
use crate::sentinels::{
    control_tokens_from_text, is_control_token, split_by_control_sequence, ANY_SENTINEL_RE,
};

#[derive(Clone, Debug)]
pub struct SpanSlice {
    pub span: FormatSpan,
    pub text: String,
}

pub fn project_translation_to_spans(
    spans: &[FormatSpan],
    source_surface: &str,
    target_surface: &str,
    nt_map: &std::collections::HashMap<String, String>,
) -> anyhow::Result<Vec<SpanSlice>> {
    if control_tokens_from_text(source_surface) != control_tokens_from_text(target_surface) {
        return Err(anyhow!("control token sequence mismatch"));
    }

    let source_parts = split_by_control_sequence(source_surface);
    let target_parts = split_by_control_sequence(target_surface);
    if source_parts.len() != target_parts.len() {
        return Err(anyhow!("control token part count mismatch"));
    }

    let mut span_slices: Vec<SpanSlice> = Vec::new();
    let mut span_idx = 0usize;

    for (src_part, tgt_part) in source_parts.into_iter().zip(target_parts.into_iter()) {
        if is_control_token(&src_part) {
            if src_part != tgt_part {
                return Err(anyhow!("control token mismatch"));
            }
            continue;
        }

        let mut block_spans: Vec<FormatSpan> = Vec::new();
        let mut block_src_len = 0usize;
        while span_idx < spans.len() && block_src_len < src_part.len() {
            let sp = spans[span_idx].clone();
            block_src_len += sp.source_text.len();
            block_spans.push(sp);
            span_idx += 1;
        }

        if block_spans.is_empty() {
            if !tgt_part.trim().is_empty() {
                return Err(anyhow!("translated text exists for empty source block"));
            }
            continue;
        }

        let tgt_units = unitize(&tgt_part);
        let total_plain = count_plain_units(&tgt_units);
        let weights: Vec<usize> = block_spans
            .iter()
            .map(|s| s.source_text.len().max(1))
            .collect();
        let desired = allocate_plain_counts(total_plain, &weights);

        let mut slices_units: Vec<Vec<String>> = vec![Vec::new(); block_spans.len()];
        let mut current_span = 0usize;
        let mut current_plain = 0usize;

        for unit in tgt_units {
            slices_units[current_span].push(unit.clone());
            if !is_sentinel(&unit) {
                current_plain += 1;
            }
            if current_span + 1 < block_spans.len() && current_plain >= desired[current_span] {
                current_span += 1;
                current_plain = 0;
            }
        }

        for (sp, units) in block_spans.into_iter().zip(slices_units.into_iter()) {
            let frozen = units.concat();
            let unfrozen = unfreeze_text(&frozen, nt_map);
            span_slices.push(SpanSlice {
                span: sp,
                text: unfrozen,
            });
        }
    }

    if span_idx != spans.len() {
        let remaining = &spans[span_idx..];
        if remaining.iter().any(|s| !s.source_text.trim().is_empty()) {
            return Err(anyhow!("span coverage mismatch"));
        }
        for sp in remaining {
            span_slices.push(SpanSlice {
                span: sp.clone(),
                text: String::new(),
            });
        }
    }

    Ok(span_slices)
}

pub fn distribute_span_text_to_nodes(span: &FormatSpan, text: &str) -> Vec<(TextNodeRef, String)> {
    if span.node_refs.is_empty() {
        return vec![];
    }
    if span.node_refs.len() == 1 {
        return vec![(span.node_refs[0].clone(), text.to_string())];
    }
    let weights: Vec<usize> = span
        .node_refs
        .iter()
        .map(|n| n.original_text.len().max(1))
        .collect();
    let units: Vec<char> = text.chars().collect();
    let desired = allocate_plain_counts(units.len(), &weights);
    let mut out: Vec<(TextNodeRef, String)> = Vec::new();

    let mut boundaries: Vec<usize> = Vec::new();
    let mut pos = 0usize;
    for count in desired.iter().take(span.node_refs.len().saturating_sub(1)) {
        pos = (pos + *count).min(units.len());
        boundaries.push(pos);
    }

    // Boundary smoothing to avoid splitting ASCII words/numbers across XML nodes.
    // This is best-effort and keeps monotonic boundaries.
    fn is_ascii_word_char(ch: char) -> bool {
        ch.is_ascii_alphanumeric()
    }
    fn boundary_quality(units: &[char], pos: usize) -> (bool, bool) {
        if pos == 0 || pos >= units.len() {
            return (true, true);
        }
        let left = units[pos - 1];
        let right = units[pos];
        let ok = !(is_ascii_word_char(left) && is_ascii_word_char(right));
        let whitespace = left.is_whitespace() || right.is_whitespace();
        (ok, whitespace)
    }

    let window = 12usize;
    let mut prev_b = 0usize;
    for b in &mut boundaries {
        let orig = *b;
        let start = orig.saturating_sub(window);
        let end = (orig + window).min(units.len());

        let mut best = orig;
        let mut best_dist = usize::MAX;
        let mut best_ws = false;
        for cand in start..=end {
            if cand < prev_b || cand > units.len() {
                continue;
            }
            let (ok, ws) = boundary_quality(&units, cand);
            if !ok {
                continue;
            }
            let dist = orig.abs_diff(cand);
            if dist < best_dist || (dist == best_dist && ws && !best_ws) {
                best = cand;
                best_dist = dist;
                best_ws = ws;
            }
        }
        *b = best;
        prev_b = best;
    }

    let mut start_idx = 0usize;
    for (i, nr) in span.node_refs.iter().cloned().enumerate() {
        let end_idx = if i < boundaries.len() {
            boundaries[i].min(units.len())
        } else {
            units.len()
        };
        let piece: String = units[start_idx..end_idx].iter().collect();
        out.push((nr, piece));
        start_idx = end_idx;
    }
    out
}

fn is_sentinel(unit: &str) -> bool {
    ANY_SENTINEL_RE.is_match(unit)
}

fn unitize(text: &str) -> Vec<String> {
    let mut units: Vec<String> = Vec::new();
    let mut pos = 0usize;
    for m in ANY_SENTINEL_RE.find_iter(text) {
        let seg = &text[pos..m.start()];
        units.extend(seg.chars().map(|c| c.to_string()));
        units.push(m.as_str().to_string());
        pos = m.end();
    }
    let tail = &text[pos..];
    units.extend(tail.chars().map(|c| c.to_string()));
    units
}

fn count_plain_units(units: &[String]) -> usize {
    units.iter().filter(|u| !is_sentinel(u)).count()
}

fn allocate_plain_counts(total_plain: usize, weights: &[usize]) -> Vec<usize> {
    if weights.is_empty() {
        return vec![];
    }
    if total_plain == 0 {
        return vec![0; weights.len()];
    }
    let total_w: usize = weights.iter().sum();
    if total_w == 0 {
        let base = total_plain / weights.len();
        let mut out = vec![base; weights.len()];
        let used: usize = out.iter().sum();
        if let Some(last) = out.last_mut() {
            *last += total_plain - used;
        }
        return out;
    }

    let raw: Vec<f64> = weights
        .iter()
        .map(|w| (total_plain as f64) * (*w as f64) / (total_w as f64))
        .collect();
    let mut floored: Vec<usize> = raw.iter().map(|x| x.floor() as usize).collect();
    let mut remain = total_plain.saturating_sub(floored.iter().sum());

    let mut frac: Vec<(f64, usize)> = raw
        .iter()
        .enumerate()
        .map(|(i, x)| (x - (floored[i] as f64), i))
        .collect();
    frac.sort_by(|a, b| b.0.partial_cmp(&a.0).unwrap_or(std::cmp::Ordering::Equal));
    let len = frac.len().max(1);
    let mut k = 0usize;
    while remain > 0 {
        let idx = frac[k % len].1;
        floored[idx] += 1;
        remain -= 1;
        k += 1;
    }
    floored
}
